### DISCUSSION TITLE
libtool: error: unknown tag cxx

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16319] [Q&A Getting Started] [answered]
Hell,

I try to install Moose on a work station. Everything goes well.

But when  
```
cd ~/projects/moose/test
make -j 4
```
It shows : **libtool: error: unknown tag cxx**

**Comment by LiuPengPeter** [comment]
Could anyone help?
It could run simulation, but will always show that problem. Will it affect?

**Comment by milljm** [comment]
What operating system are you using? Can you post some of the context of the error?

**Reply by LiuPengPeter** [reply]
Thnak you for your help!

It shows taht:

(moose) lizhen1@dlut:/Project/big_cat$ make > a.txt
libtool: error: ignoring unknown tag CXX
(moose) lizhen1@dlut:/Project/big_cat$ make
Checking if header needs updating: /home/lizhen1/Project/moose/framework/include/base/MooseRevision.h...
Rebuilding symlinks in /home/lizhen1/Project/big_cat/build/header_symlinks
Creating Unity Directory /home/lizhen1/Project/big_cat/build/unity_src
Creating Unity /home/lizhen1/Project/big_cat/build/unity_src/kernels_Unity.C
Compiling C++ (in opt mode) /home/lizhen1/Project/big_cat/build/unity_src/kernels_Unity.C...
libtool: error: ignoring unknown tag CXX
Linking Library /home/lizhen1/Project/big_cat/lib/libbig_cat-opt.la...
libtool: error: ignoring unknown tag CXX
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
/home/lizhen1/miniconda3/envs/moose/libmesh/contrib/bin/libtool: 9009: /usr/bin/grep: not found
Linking Executable /home/lizhen1/Project/big_cat/big_cat-opt...
libtool: error: ignoring unknown tag CXX

**Comment by milljm** [comment]
@lindsayad  /usr/bin/grep is also hard coded?

Okay, so this feels like a moose-libmesh issue. There is a similar discussion going on about contrib/bin/libtool in moose-libmesh #16123.

I think your only option will be to build libMesh yourself, and not use the one supplied by us. Please do the following:

```bash
conda deactivate
conda env remove -n moose
conda create -n moose moose-petsc moose-tools
conda activate moose
```

The above basically re-installs everything needed to build moose _except_ libMesh. So we will need to build that manually. But first, let's clean the moose repository, just in case the previous failed build left any lingering issues:

```bash
cd /home/lizhen1/Project/moose
git clean -xfd
```

Once moose is clean, we can try to build libMesh:

```bash
export MOOSE_JOBS=4
cd /home/lizhen1/Project/moose
scripts/update_and_rebuild_libmesh.sh
```

Hopefully, you can now build your application:

```bash
cd /home/lizhen1/Project/big_cat
make -j 4
```

**Reply by lindsayad** [reply]
> @lindsayad /usr/bin/grep is also hard coded?

Yes. I think we should build our next libmesh package on a non-meowbuild machine so that we can reproduce the grep and sed (and probably other things) paths that we had before. We're not going to have a real fix until we decide what we want to use for configuration, e.g. cmake or autotools.

**Reply by milljm** [reply]
I doubt meowbuild has anything to do with it... meowbuild is running the same singularity image as buildq, mobuild, hpcbuild, etc. Which is/was CentOS 7.7.

**Reply by lindsayad** [reply]
I dunno. You said everything was fine until this most recent libmesh package...and wasn't that the first time we used meowbuild to create the package?

**Reply by milljm** [reply]
Maybe it's the difference between CentOS 7.6 and 7.7? I was wrong in saying they are using the same image. Turns out meowbuild is using CentOS-7.7 (and now 8), while the rest are using 7.6.

Can't use 7.6 or 7.7 (valgrind issues as you discovered). So I stepped it up to 8.

I can attempt to manually build our moose-libmesh package using 8, see if that works. If not... not sure stepping backwards is the right solution.

**Reply by lindsayad** [reply]
You're right. Stepping back is not the right solution. I would say that if we're unable to get `/bin/sed, /bin/grep` with CentOS 8, then we should probably modify our conda instructions to tell the user to build libMesh...until we get an independent MOOSE configuration

**Reply by milljm** [reply]
I have an issue building libMesh on CentOS 8 using Conda supplied compilers. I have a sample PR busy working on all the moose- packages.

https://civet.inl.gov/job/637387/

I suspect it will make it, and then fail on moose-libmesh. It feels like a CentOS 8 issue. But I am not sure how to fix it.

**Reply by milljm** [reply]
Attaching the log generated by `conda build .` while in moose/conda/libmesh

[libmesh.txt](https://github.com/idaholab/moose/files/5625397/libmesh.txt)

**Comment by dschwen** [comment]
Can we (@milljm) not patch the generated `libtool` scripts? The shell command paths that cause problems are assigned to variables at the top of `libtool` (`SED="/bin/sed"`). Wouldn't it be possible to change this to `SED="sed"`?

**Reply by milljm** [reply]
And for grep as well? As the OP ran into.

**Reply by dschwen** [reply]
Yes, of course

**Comment by milljm** [comment, marked as answer]
This was fixed, along with /usr/bin/sed early this morning. Please perform the following:

```bash
conda update --all
```

And try to build again.

**Reply by mangerij** [reply]
Very cool. Thanks Jason
